### DISCUSSION TITLE
Refactor Funix

### DISCUSSION THREAD

**Discussion by Konoyoru435** [TexteaInc/funix#52] [Ideas]
# Why

Funix works fine now, for example [MFlux](https://github.com/Yazawazi/MFlux). Custom Widgets, Themes & Layout, all these features sound great. But to be honest, the code behind it is highly coupled and confusing.

# Backend

In the backend, I am thinking that we could use class/object decorators to solve the global variables like the default theme, and split some functions into separate modules. We can try to find a more elegant way to get/compare the types of parameters instead of multiple nested if-else.

# Frontend

I tried to split the [logic in `Array.Map` (`ObjectFieldExtendedTemplate`)](https://github.com/TexteaInc/funix/blob/d7c69cbf0185d736a75ebb199ee89cd682d8e344/frontend/src/components/TexteaFunction/ObjectFieldExtendedTemplate.tsx#L73L449) but ended up making a [commit](https://github.com/TexteaInc/funix/commit/d9afde7e00845561a9d1c4b64cda55fa14aaeb87) worse than it was due to the headache of the if-else, [push (for example)](https://github.com/TexteaInc/funix/blob/d9afde7e00845561a9d1c4b64cda55fa14aaeb87/frontend/src/components/TexteaFunction/ObjectFieldExtendedTemplate.tsx#L319)... 

We also need to split components like `Sheet` in `ObjectFieldExtendedTemplate`. And think of a more elegant way to solve the complex and multiple nested if-else.

# When

Thinking about it, the chart, matplotlib support that will be introduced in TEP 9, and the widgets that will be needed in the future, might be fatiguing for future developers/maintainers to make changes at that time. We can continue to maintain Funix, finish TEP 9, and then start another repo or branch to maintain Funix v2.

I apologize for the negativity that appears in the article, and I mean no offense to anyone.

**Comment by forrestbao** [comment]
I totally agree. Thanks for your passion to make Funix better. 
Let's refactorize it after the release, which should happen very soon.
